### PR TITLE
Update `updatedAtTime` only on sending a message instead of receiving it (fixes #1248)

### LINKED ISSUE #1248: JetBrains: Opening a chat from chat history triggers same chat and it is showing as recently asked
### Cody Version

v5.5.1-nightly

### IDE Information

IntelliJ IDEA 2023.3.4 (Community Edition)
Build #IC-233.14475.28, built on February 13, 2024
Runtime version: 17.0.10+1-b1087.17 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11.0
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 12
Registry:
ide.experimental.ui=true
Non-Bundled Plugins:
PythonCore (233.14475.9)
com.sourcegraph.jetbrains (5.5.1-nightly)
Kotlin: 233.14475.28-IJ

### Describe the bug

Platform: Windows

**Steps to reproduce:**

1. Have Cody plugin installed in your JB PyCharm editor
2. Login to Cody with Free/Pro/Enterprise user
3. Open Cody chat window and navigate to "Chat History"
4. Open any chat from chat history and navigate to "Chat History" again

### Observed behavior:
Chat loaded from history is triggering same chat and it is showing on top as recently asked 

### Expected behavior

Chat should not be bumped to top when loaded from history.

### Additional context

**Video link:**

https://drive.google.com/file/d/1L40H7dJNh_hhai7KU8IsqxIcrAHYKcfK/view?usp=drive_link

### PR DESCRIPTION
**Pull request by mkondratek**
Fixes #1248.

## Test plan

See the issue for specific steps.

1. Having two chats in the Chat History panel.
2. Restart IDE.
3. Go to Chat History panel and restore the older one.
4. Go to Chat History

Expected:
The restored chat remains as the last one in the list (it is not bumped to the top).

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/history/ChatHistoryPanel.kt [hunk 1]
```diff
@@ -41,7 +41,7 @@ import javax.swing.tree.DefaultTreeModel
 import javax.swing.tree.TreePath
 import javax.swing.tree.TreeSelectionModel
 
-class HistoryTree(
+class ChatHistoryPanel(
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/history/ChatHistoryPanel.kt:44]: NTOE: I piggyback this change.

This class is a panel actually, it extends `SimpleToolWindowPanel`.

File: src/main/kotlin/com/sourcegraph/cody/history/HistoryService.kt [hunk 2]
```diff
@@ -33,10 +33,14 @@ class HistoryService(private val project: Project) :
...
-  fun updateChatMessages(internalId: String, chatMessages: List<ChatMessage>) {
+  fun updateChatMessages(
+      internalId: String,
+      chatMessages: List<ChatMessage>,
+      shouldUpdateTime: Boolean
+  ) {
     val found = getOrCreateChat(internalId)
     found.messages = chatMessages.map(::convertToMessageState).toMutableList()
-    if (chatMessages.lastOrNull()?.speaker == Speaker.HUMAN) {
+    if (shouldUpdateTime) {
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/history/HistoryService.kt]: I get all that changes, and they are correct, but I do not find that variable passing particularly clean.
I was thinking if we can simplify it somehow and I get one idea:
Shouldn't we update only, if `found.messages` count increased compared to what already is in the history? What would be one-line local check in this method.

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/history/HistoryService.kt]: oh, sounds good!

File: src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt [hunk 3]
```diff
@@ -253,6 +253,7 @@ private constructor(
 
           ChatMessage(speaker = parsed, message.text)
         }
+    this.messages.addAll(chatMessages)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt]: This is needed and reasonable too. Let's add all these message in advance. `addMessageAtIndex` will not even execute in `receiveMessage` as we check the condition: `sessionMessage != incomingMessage`.

File: src/main/kotlin/com/sourcegraph/cody/history/ChatHistoryPanel.kt [hunk 4]
```diff
@@ -140,7 +140,9 @@ class HistoryTree(
       }
     } else {
       val currentPeriodText = DurationGroupFormatter.format(chat.getUpdatedTimeAt())
-      val currentPeriod = root.periods().find { it.periodText == currentPeriodText } ?: return
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/history/ChatHistoryPanel.kt:143]: this was a bug 🐛

**Review by pkukielka — APPROVED** [review]
LGTM

**Pull request merged**